### DISCUSSION TITLE
Extend DataGrid sort functionality

### DISCUSSION THREAD

**Discussion by ehsangfl** [microsoft/fluentui-blazor#283] [Ideas]
### 🙋 Feature proposal
We would like to extend DataGrid sort capabilities with new features
1. The user can sort by multi columns
    Currently, the user can only sort by a single column. we add the capability to DataGrid so that a user can sort on multiple columns, where every column can be sorted ascending or descending. The order the user selects the columns in determines the sort order.
2. Add the embedded sort capability to `TemplateColumn` without the need of extra code
    The `TemplateColumn` is a widely used column and we think an embedded sorting feature would be very useful. We want to prevent a developer needs to add code to add sort capability for every column in the application.
3. The end-user can remove the sorting and return to the original order of the data
4. The developer can easily create custom columns and DataGrid can apply sorting to these custom columns
5. Column sort capability should not be dependent on the displayed property of a column. The developer can display a property (for example display icon) and then sort by different property. 

### 💁 Possible Solution

1. Add an interface with following structure
    1- a property to define sort property of model.
    2- a function that can apply sort (or call infrastructure sort provider)
2. Add "Remove Sort" button to a column header. We suggest merging this with the already existing `ColumnOptions` to form a new embedded component. We can design it to look like Excel (or something alike).

![01](https://user-images.githubusercontent.com/10611397/210498796-478318b8-b302-4901-a49c-53abf93afa7d.jpg)

3. An infrastructure sort provider. which every column can use to apply sorting
4. Add a label/badge to column header to convey the order of the sorted columns to the user.
5. Remove the (unneeded) previous sort capabilities (like `GridSort`)

**Comment by vnbaaij** [comment]
We've decided to stay close to the QuickGrid code (which is being productized for .NET 8) and not include the changes described above.

**Reply by tommyvct** [reply]
So is there a way to remote the sort without reload the whole page?

**Reply by vnbaaij** [reply]
I assume you mean 'remove'. No, there is no way to do that without reload.

**Reply by tommyvct** [reply]
Yes, that's a typo on me, sorry. 

And by 'reload' you mean F5, right? Because I tried `StateHasChanged()` doesn't work.

Triggering hard reload just to remove the sort feels dumb, as from an end user respective, they will see the page disappears then came back, giving the false illusion of discontinuity.

**Comment by vnbaaij** [comment]
I think I've come up with a good solutionn for this... a right click on the sorted colum header button to clear the sort and restore it to the default/start situation. 
When a grid has not sort specified, it just goes back to its initial state. When thare was a defaut sort specified, it will be restored:

![grid-sort-remove](https://github.com/microsoft/fluentui-blazor/assets/1761079/b76a0830-4216-44e9-aa25-f1ef6671e320)

I'll create a PR for this.

**Reply by tommyvct** [reply]
Thank you! Will there be a callback or method that I can put on a button?

**Reply by vnbaaij** [reply]
Nope. A remove method is just there whenever a column is sortable. You can see how it is implemented in the PR #1826
